src/cursor/abstract_cursor.ts

```diff
@@ -757,18 +719,71 @@ function cursorIsDead(cursor: AbstractCursor): boolean {
   return !!cursorId && cursorId.isZero();
 }
 
-function cleanupCursor(cursor: AbstractCursor, callback: Callback): void {
-  if (cursor[kDocuments].length === 0) {
-    cursor[kClosed] = true;
-    cursor.emit(AbstractCursor.CLOSE);
+function cleanupCursor(
+  cursor: AbstractCursor,
+  options: { error?: AnyError | undefined; needsToEmitClosed?: boolean } | undefined,
+  callback: Callback
+): void {
+  const cursorId = cursor[kId];
+  const cursorNs = cursor[kNamespace];
+  const server = cursor[kServer];
+  const session = cursor[kSession];
+  const error = options?.error;
+  const needsToEmitClosed =
+    options?.needsToEmitClosed ||
+    (options?.needsToEmitClosed == null && cursor[kDocuments].length === 0);
+
+  if (error) {
+    if (cursor.loadBalanced && error instanceof MongoNetworkError) {
+      return completeCleanup();
+    }
   }
 
-  const session = cursor[kSession];
-  if (session && session.owner === cursor) {
-    session.endSession(callback);
-  } else {
-    callback();
+  if (cursorId == null || server == null || cursorId.isZero() || cursorNs == null) {
+    if (needsToEmitClosed) {
+      cursor[kClosed] = true;
```

Comment:
I've removed all the need for this in multiple places now. There is one thing I did not refactor as it would have changed the public `close` api is that the `skipKillCursors` option is never used that I could see, and passing `options` to `close` could be removed. However since 4.0 has already been released I didn't make this change. I created https://jira.mongodb.org/browse/NODE-3489 for this.

Reply:
Actually I had to revert this commit as it caused changes that broke cursors in regular mode in the case where a cursor was closed and there were still documents in memory to be iterated. I'll create another ticket to address that.

Reply:
FWIW, I think the real reason that `close` sets the flag immediately is because there is a slim chance that cursor state changes between the sync call to `close` and the `maybePromise` code called in its body, iirc a "tick" might go by between those two calls.

Reply:
I've created https://jira.mongodb.org/browse/NODE-3495 to address refactoring this.